internal/npm_install/npm_install.bzl

```diff
@@ -298,7 +310,51 @@ data attribute.
     ),
 })
 
-def _apply_patches(repository_ctx, patches):
+def _write_file(repository_ctx, path, contents):
+    bash_exe = repository_ctx.os.environ["BAZEL_SH"] if "BAZEL_SH" in repository_ctx.os.environ else "bash"
+    st = repository_ctx.execute(
+        [bash_exe, "-c", "echo '%s' > %s" % (contents, path)],
+        quiet = repository_ctx.attr.quiet,
+    )
+    if st.return_code:
+        fail("Error writing to file %s:\n%s%s" % (path, st.stderr, st.stdout))
+
+def _is_file(repository_ctx, path):
+    bash_exe = repository_ctx.os.environ["BAZEL_SH"] if "BAZEL_SH" in repository_ctx.os.environ else "bash"
+    st = repository_ctx.execute(
+        [bash_exe, "-c", "[ -f \"%s\" ] && echo 1 || echo 0" % path],
+        quiet = repository_ctx.attr.quiet,
+    )
+    if st.return_code:
+        fail("Error checking for existance of file %s:\n%s%s" % (path, st.stderr, st.stdout))
+    return True if st.stdout[0] == "1" else False
+
+def _apply_pre_install_patches(repository_ctx):
+    if len(repository_ctx.attr.pre_install_patches) == 0:
+        return
+    if repository_ctx.attr.symlink_node_modules:
+        fail("pre_install_patches cannot be used with symlink_node_modules enabled")
+    _apply_patches(repository_ctx, _WORKSPACE_REROOTED_PATH, repository_ctx.attr.pre_install_patches)
+
+def _apply_post_install_patches(repository_ctx):
+    if len(repository_ctx.attr.post_install_patches) == 0:
+        return
+    if repository_ctx.attr.symlink_node_modules:
+        print("\nWARNING: @%s post_install_patches with symlink_node_modules enabled will run in your workspace and potentially modify source files" % repository_ctx.name)
```

Comment:
since we are close to 4.0, could we simplify the allowed configurations a bit and make it a breaking change that disallows this behavior instead? I think we only introduced pre/post_install_patches for the uncommon use case of needing a patch installed for the bazel workflow while leaving the npm workflow untouched, right?

Reply:
Yes, it was intended to makes bazel-only workflow patches if and when supporting dual bazel & legacy builds.

The post_install_patches can also be used as a patch-package alternative but it gets complicated when symlink_node_modules is True

Reply:
Currently, using post_install_patches or pre_install_patches with symlink_node_modules True just fails without a good error message. This PR allows post_install_patches through and disallows pre_install_patches. Neither is a breaking change. We could alternately disallow post_install_patches as well if symlink_node_modules is True tho it looks like some users would like to use it in that shape given https://github.com/bazelbuild/rules_nodejs/issues/2774